### PR TITLE
Protect GPU kernels with appropriate nemespaces.

### LINKED ISSUE #106: Protect GPU kernels with appropriate namespace(s)
Currently all specialized GPU kernels are implemented in global namespace. Consider protecting them in `ReSolve` namespace at least.

CC @kswirydo @maksud

### PR DESCRIPTION
**Pull request by pelesh**
In this PR:
- Put GPU kernels in appropriate namespaces.
- Add templates for Doxygen documentation to kernel functions (partially documented).
- Remove unused `hipVectorKernels.h` and `cudaVectorKernels.h` files.

Closes #106.

### CODE REVIEW

File: resolve/cuda/cudaKernels.cu [hunk 1]
```diff
@@ -1,372 +1,528 @@
...
-  }
 
-  if(t < 32) {
-    s_tmp1[t] += s_tmp1[t + 32];
-    s_tmp2[t] += s_tmp2[t + 32];
 
-    s_tmp1[t] += s_tmp1[t + 16];
-    s_tmp2[t] += s_tmp2[t + 16];
+    /**
+     * @brief AXPY y = y - x*alpha where alpha is [k x 1], needed in 1 and 2 synch GMRES
```
Comment (kswirydo) [inline resolve/cuda/cudaKernels.cu:121]:
```suggestion
     * @brief AXPY y = y - x*alpha where alpha is [k x 1], and x is [N x k] needed in 1 and 2 synch GMRES
```

File: resolve/cuda/cudaKernels.cu [hunk 2]
```diff
@@ -1,372 +1,528 @@
...
+     * 
+     * @tparam Tmaxk 
+     * 
+     * @param[in]  N      -
```
Comment (kswirydo) [inline resolve/cuda/cudaKernels.cu:125]:
```suggestion
     * @param[in]  N      - number of rows in x
```

File: resolve/cuda/cudaKernels.cu [hunk 3]
```diff
@@ -1,372 +1,528 @@
...
+     * @param[in]  k      -
```
Comment (kswirydo) [inline resolve/cuda/cudaKernels.cu:126]:
```suggestion
     * @param[in]  k      - number of columns in x
```

File: resolve/cuda/cudaKernels.cu [hunk 4]
```diff
@@ -1,372 +1,528 @@
...
+     * @param[in]  x_data -
```
Comment (kswirydo) [inline resolve/cuda/cudaKernels.cu:127]:
```suggestion
     * @param[in]  x_data - double array, size [N x k]
```

File: resolve/cuda/cudaKernels.cu [hunk 5]
```diff
@@ -1,372 +1,528 @@
...
+     * @param[out] y_data -
```
Comment (kswirydo) [inline resolve/cuda/cudaKernels.cu:128]:
```suggestion
     * @param[out] y_data - double array, size [N x 1]
```

File: resolve/cuda/cudaKernels.cu [hunk 6]
```diff
@@ -1,372 +1,528 @@
...
+     * @param[in]  alpha  -
```
Comment (kswirydo) [inline resolve/cuda/cudaKernels.cu:129]:
```suggestion
     * @param[in]  alpha  - doble array, size [k x 1]
```

File: resolve/cuda/cudaKernels.cu [hunk 7]
```diff
@@ -1,372 +1,528 @@
...
+        y_data[i] -= temp;
+      }
+    }
 
-    s_tmp1[t] += s_tmp1[t + 8];
-    s_tmp2[t] += s_tmp2[t + 8];
+    /**
+     * @brief Pass through matrix rows and sum each as \sum_{j=1}^m abs(a_{ij})
+     * 
+     * @param[in]  n      -
```
Comment (kswirydo) [inline resolve/cuda/cudaKernels.cu:159]:
```suggestion
     * @param[in]  n      - number of rows in the matrix.
```

File: resolve/cuda/cudaKernels.cu [hunk 8]
```diff
@@ -1,372 +1,528 @@
...
+     * @param[in]  nnz    -
```
Comment (kswirydo) [inline resolve/cuda/cudaKernels.cu:160]:
```suggestion
     * @param[in]  nnz    - number of non-zero values in the matrix
```

File: resolve/cuda/cudaKernels.cu [hunk 9]
```diff
@@ -1,372 +1,528 @@
...
+     * @param[in]  a_ia   -
```
Comment (kswirydo) [inline resolve/cuda/cudaKernels.cu:161]:
```suggestion
     * @param[in]  a_ia   - row pointers (CSR storage)
```

File: resolve/cuda/cudaKernels.cu [hunk 10]
```diff
@@ -1,372 +1,528 @@
...
+     * @param[in]  a_val  -
```
Comment (kswirydo) [inline resolve/cuda/cudaKernels.cu:162]:
```suggestion
     * @param[in]  a_val  - values (CSR storage)
```

File: resolve/cuda/cudaKernels.cu [hunk 11]
```diff
@@ -1,372 +1,528 @@
...
+     * @param[out] result -
```
Comment (kswirydo) [inline resolve/cuda/cudaKernels.cu:163]:
```suggestion
     * @param[out] result - array size [n x 1] containing sums of values in each row.
```

File: resolve/cuda/cudaKernels.cu [hunk 12]
```diff
@@ -1,372 +1,528 @@
...
+        idx += (blockDim.x*gridDim.x);
+      }
+    }
 
-    s_tmp1[t] += s_tmp1[t + 4];
-    s_tmp2[t] += s_tmp2[t + 4];
+    /**
+     * @brief For count sketch in sketching random method
+     * 
+     * @param[in]  n      -
```
Comment (kswirydo) [inline resolve/cuda/cudaKernels.cu:185]:
```suggestion
     * @param[in]  n      - number of entries in input vector
```

File: resolve/cuda/cudaKernels.cu [hunk 13]
```diff
@@ -1,372 +1,528 @@
...
+     * @param[in]  k      -
```
Comment (kswirydo) [inline resolve/cuda/cudaKernels.cu:186]:
```suggestion
     * @param[in]  k      - number of entries in output vector (k < n)
```

File: resolve/cuda/cudaKernels.cu [hunk 14]
```diff
@@ -1,372 +1,528 @@
...
+     * @param[in]  labels -
```
Comment (kswirydo) [inline resolve/cuda/cudaKernels.cu:187]:
```suggestion
     * @param[in]  labels - array size [n x 1] containing integers from 0 to k-1, assigned randomly.
```

File: resolve/cuda/cudaKernels.cu [hunk 15]
```diff
@@ -1,372 +1,528 @@
...
+     * @param[in]  flip   -
+     * @param[in]  input  -
```
Comment (kswirydo) [inline resolve/cuda/cudaKernels.cu:189]:
```suggestion
     * @param[in]  input  - input vector, size [n x 1] 
```

File: resolve/cuda/cudaKernels.cu [hunk 16]
```diff
@@ -1,372 +1,528 @@
...
-    s_tmp2[t] += s_tmp2[t + 4];
+    /**
+     * @brief For count sketch in sketching random method
+     * 
+     * @param[in]  n      -
+     * @param[in]  k      -
+     * @param[in]  labels -
+     * @param[in]  flip   -
+     * @param[in]  input  - 
+     * @param[out] output -
```
Comment (kswirydo) [inline resolve/cuda/cudaKernels.cu:190]:
```suggestion
     * @param[out] output - output vector, size [k x 1] (this vector must be allocated and initialized with `0`s prior to calling the kernel).
```

File: resolve/cuda/cudaKernels.cu [hunk 17]
```diff
@@ -1,372 +1,528 @@
...
+                             real_type* y)
+    {
+      index_type idx = blockIdx.x * blockDim.x + threadIdx.x; 
+
+      while (idx < n) {
+
+        if (D[idx] == 1) {
+          y[idx] = x[idx];
+        } else {
+          y[idx] = (-1.0)*x[idx];
```
Comment (kswirydo) [inline resolve/cuda/cudaKernels.cu:250]:
```suggestion
          y[idx] = (-1.0) * x[idx];
```

File: resolve/cuda/cudaKernels.cu [hunk 18]
```diff
@@ -1,372 +1,528 @@
...
+     * @param[in]  a_val  -
+     * @param[out] result -
+     */
+    __global__ void matrixInfNormPart1(const index_type n, 
+                                       const index_type nnz, 
+                                       const index_type* a_ia,
+                                       const real_type* a_val, 
+                                       real_type* result)
+    {
+      index_type idx = blockIdx.x*blockDim.x + threadIdx.x;
```
[the remark this answers is not in the document]
Comment (pelesh) [reply resolve/cuda/cudaKernels.cu]: Fixed!

**Pull request merged**